### PR TITLE
handle batch EVM requests

### PR DESCRIPTION
**Pull request by pirtleshell**
Adds support for batch EVM queries.

Batch queries have bodies with a list of `EVMRPCRequestEnvelope`s instead of just one. Currently, these fail to decode and then skip the cache, skip metrics, and just get forwarded to the appropriate backend as-is.

This PR handles decoding the batch requests and supports getting responses to their sub-requests from the cache. To do this, it adds two new middlewares:
* `DecodeRequestMiddleware` - a rewrite of `RequestLoggingMiddleware`. now attempts to parse incoming request as a `[]*EVMRPCRequestEnvelope` if it fails to parse as `*EVMRPCRequestEnvelope`. Forks the middleware sequence if the request is a batch
* `BatchProcessingMiddleware`  - pulls the batch request out of the context, separates into individual requests that can leverage response caching (and metrics). Then, it concurrently makes the individual requests through the same middleware sequence as the current single request processing. All the individual responses are combined back into a single response to the client.

For minimal changes to existing architecture, the BatchProcessingMiddleware just makes use of the exact same middleware sequence as the handling of single requests. Optimizations can be made in the future to recombine sub-requests that will be directed to the same backend if performance is a problem (we have metrics to now to know!). I excluded those changes for now to minimize potential impact to existing single request response flow.

~~Opening for code review, will not merge until architecture docs have been updated.~~

### CODE REVIEW

File: routines/metric_partitioning_test.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ func TestE2ETestMetricPartitioningRoutinePrefillsExpectedPartitionsAfterStartupD
 	// prepare
 	time.Sleep(time.Duration(MetricPartitioningRoutineDelayFirstRunSeconds) * time.Second)
 
-	expectedPartitions, err := partitionsForPeriod(time.Now(), int(configuredPrefillDays))
+	expectedPartitions, err := partitionsForPeriod(time.Now().UTC(), int(configuredPrefillDays))
```
Comment (galxy25) [inline routines/metric_partitioning_test.go:37]: universal time for the global win

File: service/batchmdw/fake_response_writer.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package batchmdw
+
+import (
+	"bytes"
+	"net/http"
+)
+
+// fakeResponseWriter is a custom implementation of http.ResponseWriter that writes all content
+// to a buffer.
+type fakeResponseWriter struct {
```
Comment (galxy25) [inline service/batchmdw/fake_response_writer.go:10]: `bufferedResponseWriter` a la https://pkg.go.dev/bufio?

File: service/cachemdw/cache.go [hunk 3]
```diff
@@ -80,6 +80,13 @@ func IsCacheable(
 	logger *logging.ServiceLogger,
 	req *decode.EVMRPCRequestEnvelope,
 ) bool {
+	// TODO: technically, we _could_ cache the "invalid request" response for `null` requests...
+	// however, doing so will may result in different than expected responses for batch requests
```
Comment (galxy25) [inline service/cachemdw/cache.go:84]: `may will`

File: service/batchmdw/batch_processor.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		status:    http.StatusOK,
+		mu:        sync.Mutex{},
+	}
+}
+
+// RequestAndServe concurrently sends each request to the underlying handler
+// Responses are then collated into a JSON array and written to the ResponseWriter
+func (bp *BatchProcessor) RequestAndServe(w http.ResponseWriter) error {
+	wg := sync.WaitGroup{}
+	for i, r := range bp.requests {
```
Comment (galxy25) [inline service/batchmdw/batch_processor.go:41]: unlikely but potential DDOS vector, might make sense to pick some reasonably high limit (100, 1000?) configured via env for how much request amplification a single batched request can place on the backend fleet

**Review by galxy25 — APPROVED** [review]

File: service/batchmdw/batch_processor.go (hunk 4 above)
Comment (pirtleshell) [reply service/batchmdw/batch_processor.go:41]: great point, didn't think through the fact that this removes _any_ batch limit that may be imposed by the actual nodes 😅 🙏

Comment (pirtleshell) [reply service/batchmdw/batch_processor.go:41]: i updated to have the batch processing middleware take and respect a maximum batch size. if it gets a request more sub-requests than that, it responds 413.

File: service/batchmdw/fake_response_writer.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	// body is the response body for the current request
+	body *bytes.Buffer
+	// header is the response headers for the current request
+	header http.Header
+	// onErrStatus is a method for handling non-OK status responses
+	onErrStatus func(status int)
```
[the remark this answers is not in the document]
Comment (pirtleshell) [reply service/batchmdw/fake_response_writer.go:16]: do you have an example? i can pass the response body to the err handler as well. is that what you mean? though, for error statuses, i don't think the response will have anything written to it at the time of WriteHeader being called.

the response writer doesn't have access to the request, only the response data written by http handler.

**Pull request merged**